pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -268,21 +276,26 @@ func (configgen *ConfigGeneratorImpl) buildGatewayTCPBasedFilterChains(
 				// we are building a network filter chain (no http connection manager) for this filter chain
 				// For network filters such as mysql, mongo, etc., we need the filter codec upfront. Data from this
 				// codec is used by RBAC later.
-				var tcpAuthFilters []*listener.Filter
-				tcpAuthFilters = append(tcpAuthFilters, xdsfilters.IstioNetworkAuthenticationFilter)
-				tcpAuthFilters = append(tcpAuthFilters, builder.authzCustomBuilder.BuildTCP()...)
-				tcpAuthFilters = append(tcpAuthFilters, builder.authzBuilder.BuildTCP()...)
+				var tcpAuthAndWasmFilters []*listener.Filter
+				tcpAuthAndWasmFilters = append(tcpAuthAndWasmFilters, xdsfilters.IstioNetworkAuthenticationFilter)
+				tcpAuthAndWasmFilters = extension.PopAppendNetwork(tcpAuthAndWasmFilters, wasm, extensions.PluginPhase_AUTHN)
+				tcpAuthAndWasmFilters = append(tcpAuthAndWasmFilters, builder.authzCustomBuilder.BuildTCP()...)
+				tcpAuthAndWasmFilters = append(tcpAuthAndWasmFilters, builder.authzBuilder.BuildTCP()...)
+				tcpAuthAndWasmFilters = extension.PopAppendNetwork(tcpAuthAndWasmFilters, wasm, extensions.PluginPhase_AUTHZ)
+				tcpAuthAndWasmFilters = extension.PopAppendNetwork(tcpAuthAndWasmFilters, wasm, extensions.PluginPhase_STATS)
```

Comment:
don't we need TCP stats between here?

Reply:
> Simplified the code a bit and ensured network filter order is maintained with pre built filters. PTAL.

Part of my commentary is I think the _existing_ behavior for TCP GW is wrong actually. But haven't gotten around to fixing it; for now just added tests in #46602

Reply:
> don't we need TCP stats between here?

May be. But I just added the existing filters. Do you want me to add that in this PR - should we do it when you fix it?